### PR TITLE
[FEATURE] sam_dna16 alphabet.

### PR DESCRIPTION
**Pull request by smehringer**

### CODE REVIEW

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \author Svenja Mehringer <svenja.mehringer AT fu-berlin.de>
+ * \brief Contains seqan3::sam_eq and seqan3::sam_dna16.
+ */
+
+#pragma once
+
+#include <cassert>
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: why do you need this?

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+#include <seqan3/alphabet/detail/alphabet_base.hpp>
+#include <seqan3/alphabet/nucleotide/dna15.hpp>
+
+namespace seqan3
+{
+
+/*!\brief The alphabet of an equal character '='.
+ * \ingroup nucleotide
+ * \implements seqan3::Alphabet
+ * \implements seqan3::detail::ConstexprAlphabet
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: guard this with \if ....
Also check the writable concepts...

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr sam_eq() noexcept : base_t{} {}            //!< Defaulted.
+    constexpr sam_eq(sam_eq const &) = default;             //!< Defaulted.
+    constexpr sam_eq(sam_eq &&) = default;                  //!< Defaulted.
+    constexpr sam_eq & operator=(sam_eq const &) = default; //!< Defaulted.
+    constexpr sam_eq & operator=(sam_eq &&) = default;      //!< Defaulted.
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]:
```suggestion
    constexpr sam_eq & operator=(sam_eq &&)            noexcept = default;  //!< Defaulted.
```

Alsign the last three words. Also acutally default the default constructor

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    using base_t::base_t;
+    //!\}
+};
+
+/*!\brief A 16 letter DNA alphabet, containing all IUPAC smybols minus the gap and plus an equal sign.
+ * \ingroup nucleotide
+ * \implements seqan3::NucleotideAlphabet
+ * \implements seqan3::detail::ConstexprAlphabet
+ * \implements seqan3::TriviallyCopyable
+ * \implements seqan3::StandardLayout
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: concepts see above

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+ *
+ * The alphabet may be brace initialized from the static letter members. Note that you cannot
+ * assign the alphabet by using letters of type `char`, but you instead have to use the
+ * function seqan3::sam_dna16::assign_char().
+ *
+ *\include test/snippet/alphabet/nucleotide/sam_dna16.cpp
+ */
+using sam_dna16 = alphabet_variant<dna15, sam_eq>;
+
+
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: double newline

File: test/snippet/alphabet/nucleotide/sam_dna16.cpp [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    // sam_dna16_vector foo{"ACGTTA"};
+    // sam_dna16_vector bar = "ACGTTA";
+
+    // but these do:
+    std::vector<sam_dna16> foo{"ACgtTA"_sam_dna16};
+    std::vector<sam_dna16> bar = "ACG==A"_sam_dna16;
+    auto bax = "A=GTT!"_sam_dna16;
+
+    debug_stream << foo << "\n" << bar << "\n" << bax << "\n";
+}
```
Comment (h-2) [inline test/snippet/alphabet/nucleotide/sam_dna16.cpp]: unit tests are missing

**Comment by smehringer** [comment]
@h-2 Two things:

1. I cannot make the defaulted default constructor constexpr because the compiler thinks that alphabet base is not constexpr (although it is, I checked). :thinking: 

```
n file included from /srv/public/smehringer/seqan3/test/unit/alphabet/nucleotide/sam_dna16_test.cpp:8:
/srv/public/smehringer/seqan3/include/seqan3/alphabet/nucleotide/sam_dna16.hpp:51:15: error: explicitly defaulted function ‘constexpr seqan3::sam_eq::sam_eq()’ cannot be declared as ‘constexpr’ because the implicit declaration is not ‘constexpr’:
     constexpr sam_eq()                           noexcept = default; //!< Defaulted.
               ^~~~~~
In file included from /srv/public/smehringer/seqan3/include/seqan3/alphabet/composite/alphabet_variant.hpp:26,
                 from /srv/public/smehringer/seqan3/include/seqan3/alphabet/nucleotide/sam_dna16.hpp:15,
                 from /srv/public/smehringer/seqan3/test/unit/alphabet/nucleotide/sam_dna16_test.cpp:8:
/srv/public/smehringer/seqan3/include/seqan3/alphabet/detail/alphabet_base.hpp:251:15: note: defaulted constructor calls non-‘constexpr’ ‘seqan3::alphabet_base<derived_type, 1, char_t>::alphabet_base() [with derived_type = seqan3::sam_eq; char_t = char]’
     constexpr alphabet_base() noexcept = default;                         //!< Defaulted.
               ^~~~~~~~~~~~~
/srv/public/smehringer/seqan3/include/seqan3/alphabet/detail/alphabet_base.hpp:251:15: note: ‘seqan3::alphabet_base<derived_type, 1, char_t>::alphabet_base() [with derived_type = seqan3::sam_eq; char_t = char]’ is not usable as a ‘constexpr’ function because:
/srv/public/smehringer/seqan3/include/seqan3/alphabet/detail/alphabet_base.hpp:335:10: note: defaulted default constructor does not initialize ‘bool seqan3::alphabet_base<seqan3::sam_eq, 1, char>::_bug_workaround’
     bool _bug_workaround; // See GCC Bug-Report: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87113
          ^~~~~~~~~~~~~~~
alphabet/nucleotide/CMakeFiles/sam_dna16_test.dir/build.make:62: die Regel für Ziel „alphabet/nucleotide/CMakeFiles/sam_dna16_test.dir/sam_dna16_test.cpp.o“ scheiterte
make[3]: *** [alphabet/nucleotide/CMakeFiles/sam_dna16_test.dir/sam_dna16_test.cpp.o] Fehler 1
CMakeFiles/Makefile2:3117: die Regel für Ziel „alphabet/nucleotide/CMakeFiles/sam_dna16_test.dir/all“ scheiterte
make[2]: *** [alphabet/nucleotide/CMakeFiles/sam_dna16_test.dir/all] Fehler 2
CMakeFiles/Makefile2:3129: die Regel für Ziel „alphabet/nucleotide/CMakeFiles/sam_dna16_test.dir/rule“ scheiterte
make[1]: *** [alphabet/nucleotide/CMakeFiles/sam_dna16_test.dir/rule] Fehler 2
Makefile:1039: die Regel für Ziel „sam_dna16_test“ scheiterte
make: *** [sam_dna16_test] Fehler 2
make sam_dna16_test  5,26s user 0,30s system 95% cpu 5,812 total
```

2. We need to discuss how to handle complement on alphabet_variants. Right now my sam_dna16 is no nucleotide but I want it to be... (besides sam_eq '=' does not have a complement so far, we could make a representation for this?)

**Comment by h-2** [comment]
> I cannot make the defaulted default constructor constexpr

You need to change that line in `alphabet_base` to be:
```
bool _bug_workaround{};
```

While you are at it, can you wrap that ugliness in the following:
```
#if defined(__GNUC__) && (__GNUC__ == 7)
```
It as been fixed in GCC8 and GCC9.

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    constexpr sam_eq(sam_eq &&)                  noexcept = default; //!< Defaulted.
+    constexpr sam_eq & operator=(sam_eq const &) noexcept = default; //!< Defaulted.
+    constexpr sam_eq & operator=(sam_eq &&)      noexcept = default; //!< Defaulted.
+    ~sam_eq()                                    noexcept = default; //!< Defaulted.
+
+    using base_t::base_t;
+    //!\}
+};
+
+/*!\brief A 16 letter DNA alphabet, containing all IUPAC smybols minus the gap and plus an equal sign.
```
[the remark this answers is not in the document]
Comment (joergi-w) [reply include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: Does IUPAC contain the gap?? 😱

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+ * \implements seqan3::TriviallyCopyable
+ * \implements seqan3::StandardLayout
+ * \implements seqan3::WritableAlphabet
+ * \if DEV \implements seqan3::detail::ConstexprAlphabet \endif
+ * \if DEV \implements seqan3::detail::WritableConstexprAlphabet \endif
+ *
+ * The alphabet always has the same value ('=') which represents that the base
+ * at this position is equal to the aligned reference base.
+ */
+
```
Comment (joergi-w) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: We usually do not have empty lines between documentation and class definition.

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+constexpr sam_dna16 operator""_sam_dna16(char const c) noexcept
+{
+    return sam_dna16{}.assign_char(c);
+}
+
+/*!\brief The seqan3::sam_dna16 string literal.
+ * \relates seqan3::sam_dna16
+ * \returns seqan3::sam_dna16_vector
+ *
+ * You can use this string literal to easily assign to sam_dna16_vector:
```
Comment (joergi-w) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: seqan3::sam_dna16_vector

File: test/snippet/alphabet/nucleotide/sam_dna16.cpp [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+    my_letter.assign_char('F'); // unknown characters are implicitly converted to N.
+    debug_stream << my_letter << std::endl; // "N";
+
+    //! [operator""_sam_dna16]
+    // these don't work:
+    // sam_dna16_vector foo{"ACGTTA"};
+    // sam_dna16_vector bar = "ACGTTA";
+
+    // but these do:
+    std::vector<sam_dna16> foo{"ACgtTA"_sam_dna16};
```
Comment (joergi-w) [inline test/snippet/alphabet/nucleotide/sam_dna16.cpp]: why don't you use `sam_dna16_vector foo` as in the negative examples above?

File: test/unit/alphabet/nucleotide/sam_dna16_test.cpp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+    EXPECT_EQ(to_char(sam_dna16{}.assign_char('K')), 'K');
+    EXPECT_EQ(to_char(sam_dna16{}.assign_char('M')), 'M');
+    EXPECT_EQ(to_char(sam_dna16{}.assign_char('B')), 'B');
+    EXPECT_EQ(to_char(sam_dna16{}.assign_char('D')), 'D');
+    EXPECT_EQ(to_char(sam_dna16{}.assign_char('H')), 'H');
+    EXPECT_EQ(to_char(sam_dna16{}.assign_char('V')), 'V');
+
+    EXPECT_EQ(to_char(sam_dna16{}.assign_char('=')), '=');
+
+    EXPECT_EQ(to_char(sam_dna16{}.assign_char('N')), 'N');
```
Comment (joergi-w) [inline test/unit/alphabet/nucleotide/sam_dna16_test.cpp]: for most of them you can use a loop
`for (char chr : std::string{"ACGTRY..."})`

**Comment by smehringer** [comment]
> Sorry, I do not like the design of defining an alphabet just for the equality sign. It is only useful in one very specific context (as extension of dna15) and in the API it will be visible on the same level as our important dna and aa alphabets.
For dna5 and rna5 we did not define a new alphabet over 'N' just to extend dna4 and rna4, respectively, but rather created them new. In the same way I would prefer to define sam_dna16. 

Hm, I see your point but the reasoning behind that decision was that I do not want to copy and paste dna15 again just to add the char `=`, but use a union (alphabet_variant) of dna15 and the `equal` char instead.. I don't see a way to do both, avoiding copy and paste from dna15 and avoiding an extra equality alphabet :(

**Comment by joergi-w** [comment]
Essentially you would just define new `rank_to_char` and `char_to_rank` tables for sam_dna16, which are even slightly different from dna15, so I don't see a large copy'n'paste there. Also debugging and understanding the functionality of the alphabet would be easier if it was independent of dna15.

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (smehringer) [reply include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: yup: https://www.bioinformatics.org/sms2/iupac.html

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+namespace seqan3
+{
+
+/*!\brief A 16 letter DNA alphabet, containing all IUPAC symbols minus the gap and plus an equality sign ('=').
+ * \ingroup nucleotide
+ * \implements seqan3::WritableAlphabet
+ * \implements seqan3::NucleotideAlphabet
+ * \implements seqan3::TriviallyCopyable
+ * \implements seqan3::StandardLayout
+ * \if DEV \implements seqan3::detail::ConstexprAlphabet \endif
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]:
```suggestion
 * \if DEV \implements seqan3::detail::WritableConstexprAlphabet \endif
```

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+/*!\brief The seqan3::sam_dna16 string literal.
+ * \relates seqan3::sam_dna16
+ * \returns seqan3::sam_dna16_vector
+ *
+ * You can use this string literal to easily assign to seqan3::sam_dna16_vector:
+ *
+ * \snippet test/snippet/alphabet/nucleotide/sam_dna16.cpp operator""_sam_dna16
+ *
+ */
+inline sam_dna16_vector operator""_sam_dna16(char const * s, std::size_t n)
```
Comment (joergi-w) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: You can omit `std::` for size_t.

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+ * \returns seqan3::sam_dna16
+ */
+constexpr sam_dna16 operator""_sam_dna16(char const c) noexcept
+{
+    return sam_dna16{}.assign_char(c);
+}
+
+/*!\brief The seqan3::sam_dna16 string literal.
+ * \relates seqan3::sam_dna16
+ * \returns seqan3::sam_dna16_vector
```
Comment (joergi-w) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: missing \param `s` and `n`

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+// literals
+// ------------------------------------------------------------------
+
+/*!\name Literals
+ * \{
+ */
+
+/*!\brief The seqan3::sam_dna16 char literal.
+ * \relates seqan3::sam_dna16
+ * \returns seqan3::sam_dna16
```
Comment (joergi-w) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: missing \param `c`

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+/*!\brief The seqan3::sam_dna16 string literal.
+ * \relates seqan3::sam_dna16
+ * \returns seqan3::sam_dna16_vector
+ *
+ * You can use this string literal to easily assign to seqan3::sam_dna16_vector:
+ *
+ * \snippet test/snippet/alphabet/nucleotide/sam_dna16.cpp operator""_sam_dna16
+ *
```
Comment (joergi-w) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: trailing empty line

File: test/unit/alphabet/nucleotide/sam_dna16_test.cpp [hunk 17]
```diff
@@ -0,0 +1,111 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#include <seqan3/alphabet/nucleotide/sam_dna16.hpp>
+
```
Comment (joergi-w) [inline test/unit/alphabet/nucleotide/sam_dna16_test.cpp]: remove empty line

File: test/unit/alphabet/nucleotide/sam_dna16_test.cpp [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+using namespace seqan3;
+
+// ------------------------------------------------------------------
+// sam_dna16 alphabet
+// ------------------------------------------------------------------
+
+INSTANTIATE_TYPED_TEST_CASE_P(sam_dna16, alphabet, sam_dna16);
+INSTANTIATE_TYPED_TEST_CASE_P(sam_dna16, alphabet_constexpr, sam_dna16);
+
+// nucleotide test: (because the complement is not bijective for sam_dna16 we need to test in manually)
```
Comment (joergi-w) [inline test/unit/alphabet/nucleotide/sam_dna16_test.cpp]: ... test **it** manually?

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 19]
```diff
@@ -0,0 +1,191 @@
...
+ * \implements seqan3::WritableAlphabet
+ * \implements seqan3::NucleotideAlphabet
+ * \implements seqan3::TriviallyCopyable
+ * \implements seqan3::StandardLayout
+ * \if DEV \implements seqan3::detail::ConstexprAlphabet \endif
+ *
+ * \details
+ *
+ * The seqan3::sam_dna16 alphabet is used for aligned sam_dna16 sequences with the extra alphabet character
+ * represented by '=', which denotes that the aligned base is equal to the respective reference sequence.
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: The seqan3::sam_dna16 alphabet is the nucleotide alphabet used inside the SAM, BAM and CRAM formats.
It has all the letters of the seqan3::dna15 alphabet and the extra alphabet character '=' which denotes a nucleotide character *identical to the reference*. Without the context of this reference sequence, no assumptions can be made about the actual value of `=` letter.

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 20]
```diff
@@ -0,0 +1,191 @@
...
+ *
+ * Note that you can assign 'U' as a character to sam_dna16 and it will silently
+ * be converted to 'T'.
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: Lower case letters are accepted when assigning from `char`. The complement of a `=` letter is `N`.

File: include/seqan3/alphabet/nucleotide/sam_dna16.hpp [hunk 21]
```diff
@@ -0,0 +1,191 @@
...
+ *
+ * The alphabet may be brace initialized from the static letter members. Note that you cannot
+ * assign the alphabet by using letters of type `char`, but you instead have to use the
+ * function seqan3::sam_dna16::assign_char().
```
Comment (h-2) [inline include/seqan3/alphabet/nucleotide/sam_dna16.hpp]: This is redundant (and also wrong, there are no static letter members anymore). I would just remove it.

**Comment by h-2** [comment]
snippet and documentation failures...

**Pull request merged**